pkg/proxy/backend/backend_conn_mgr_test.go

```diff
@@ -683,3 +686,42 @@ func TestGracefulCloseBeforeHandshake(t *testing.T) {
 	}
 	ts.runTests(runners)
 }
+
+func TestGetBackendIO(t *testing.T) {
+	listeners := make([]net.Listener, 0, 3)
+	addrs := make([]string, 0, 3)
+	for i := 0; i < 3; i++ {
+		listener, err := net.Listen("tcp", "0.0.0.0:0")
+		require.NoError(t, err)
+		listeners = append(listeners, listener)
+		addrs = append(addrs, listener.Addr().String())
+	}
+	rt := router.NewStaticRouter(addrs)
+	badAddrs := make(map[string]struct{}, 3)
+	handler := &CustomHandshakeHandler{
+		getRouter: func(ctx ConnContext, resp *pnet.HandshakeResp) (router.Router, error) {
+			return rt, nil
+		},
+		onHandshake: func(connContext ConnContext, s string, err error) {
+			if err != nil && len(s) > 0 {
+				badAddrs[s] = struct{}{}
+			}
+		},
+	}
+	mgr := NewBackendConnManager(logger.CreateLoggerForTest(t), handler, 0, false, false)
+	auth := &Authenticator{}
+	for i := 0; i <= len(listeners); i++ {
+		io, err := mgr.getBackendIO(auth, auth, nil)
+		if i < len(listeners) {
+			require.NoError(t, err)
+			err = io.Close()
+			require.NoError(t, err)
+			err = listeners[i].Close()
+			require.NoError(t, err)
```

Comment:
Maybe use `t.Cleanup` to close listeners.

Reply:
It's not just cleanup, it's testing that the selector still returns an available listener even if some are closed.